### PR TITLE
[ticket/15257] Provide extension not enableable messages

### PR DESCRIPTION
**Pull request by senky**
PHPBB3-15257

Replaces #4863, in honor of javiexin.

I took different approach - `is_enableable` now allows to return array apart from boolean. `true` means the ext can be enabled, `false` that it cannot be enabled and should report generic reason, array means it cannot be enabled and reasons are attached.

Checklist:

- [x] Correct branch: master for new features; 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.2.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-15257

### CODE REVIEW

File: tests/extension/ext/vendor5/foo/ext.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	{
+		return array('Reason 1', 'Reason 2');
+	}
+
+	public function enable_step($old_state)
+	{
+		self::$enabled = true;
+		return self::$enabled;
+	}
+}
```
Comment (mrgoldy) [inline tests/extension/ext/vendor5/foo/ext.php]: No new line at end of file?

File: phpBB/includes/acp/acp_extensions.php [hunk 2]
```diff
@@ -1130,4 +1126,14 @@ public function output_metadata_to_template($metadata)
 			));
 		}
 	}
+
+	protected function check_is_enableable($enableable)
```
Comment (marc1706) [inline phpBB/includes/acp/acp_extensions.php]: Please add a docblock to this.

File: tests/extension/ext/vendor5/foo/composer.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	"require": {
+		"php": ">=5.3"
+	},
+	"extra": {
+		"display-name": "phpBB Bar Extension",
+		"soft-require": {
+			"phpbb/phpbb": "3.3.*@dev"
+		}
+	}
+}
```
Comment (marc1706) [inline tests/extension/ext/vendor5/foo/composer.json]: Add new line at the end of the file

File: tests/extension/ext/vendor5/foo/ext.php (hunk 1 above)
Comment (marc1706) [reply tests/extension/ext/vendor5/foo/ext.php]: Should be added

File: phpBB/includes/acp/acp_extensions.php (hunk 2 above)
Comment (VSEphpbb) [reply phpBB/includes/acp/acp_extensions.php]: Just a thought... Another approach (which I think is may be more cognitive/readable, could be like):
```php
$extension = $this->ext_manager->get_extension($ext_name);
if ($message = $extension->is_enableable() !== true)
{
	$this->show_error($message);
}

/**
* Trigger an error message for an extension that could not be enabled.
* Error message can be set by the extension.
*
* @param bool|array $message Array of reasons (set by the extension) 
* or false for default reason.
*/
public function show_error($message)
{
	if (is_array($message))
	{
		$message = implode('<br />', $message)
	}
	else if (empty($message)
	{
		$message = $this->user->lang('EXTENSION_NOT_ENABLEABLE');
	}

	trigger_error($message . adm_back_link($this->u_action), E_USER_WARNING);
}
```

**Comment by marc1706** [comment]
@VSEphpbb better now?

**Pull request merged**